hera_sim/simulate.py

```diff
@@ -0,0 +1,335 @@
+"""
+Primary interface module for hera_sim, defining a :class:`Simulator` class which provides a common API for all
+effects produced by this package.
+"""
+
+import functools
+import inspect
+import sys
+
+from pyuvdata import UVData
+
+from . import io
+from . import sigchain
+from .version import version
+
+
+def _get_model(mod, name):
+    return getattr(sys.modules["hera_sim." + mod], name)
+
+
+class _model(object):
+
+    def __init__(self, base_module=None):
+        self.base_module = base_module
+
+    def __call__(self, func, *args, **kwargs):
+        name = func.__name__
+
+        @functools.wraps(func)
+        def new_func(obj, *args, **kwargs):
+            if "model" in inspect.getargspec(func)[0]:
+                # Cases where there is a choice of model
+                model = args[0] if args else kwargs.pop("model")
+
+                # If the model is a str, get its actual callable.
+                if isinstance(model, str):
+                    if self.base_module is None:
+                        self.base_module = name[4:]  # get the bit after the "add"
+
+                    model = _get_model(self.base_module, model)
+
+                func(obj, model, **kwargs)
+
+                if not isinstance(model, str):
+                    method = model.__name__
+                else:
+                    method = model
+
+                method = "using {} ".format(method)
+            else:
+                # For cases where there is no choice of model.
+                method = ""
+                func(obj, **kwargs)
+
+            msg = "\nhera_sim v{version}: Added {component} {method_name}with kwargs: {kwargs}"
+
+            obj.data.history += msg.format(
+                version=version,
+                component="".join(name.split("_")[1:]),
+                method_name=method,
+                kwargs=kwargs,
+            )
+
+        return new_func
+
+
+class Simulator:
+    """
+    Primary interface object for hera_sim.
+
+    Produces visibility simulations with various independent sky- and instrumental-effects, and offers the resulting
+    visibilities in :class:`UVData` format
+    """
+
+    def __init__(
+        self,
+        data_filename=None,
```

Comment:
can we add a keyword argument that "whitens" the input data if provided. a use-case is that we have a `UVData` that we want to use as a template to get metadata, but we want to start over from scratch. So something like
```
uv.data_array[:] = 0.0
uv.flag_array[:] = False
uv.nsample_array[:] = 1.0
```
if we want to whiten it out...

Reply:
Hmm, interesting idea, but no the metadata are intimately tied to the data array, so I think whitening the data array should be accompanied by also whitening the flags and nsamples. If in the future we develop some need for it we can always add that in later--but for now better to not over-optimize I think.